informer/integration/test/deletionerror/filter_watcher.go

```diff
@@ -0,0 +1,53 @@
+// +build k8srequired
+
+package deletionerror
+
+import (
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/operatorkit/informer"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/watch"
+)
+
+type FilterWatcherConfig struct {
+	Watcher informer.Watcher
+}
+
+// FilterWatcher is a wrapping implementation of infomrmer.Watch used to filter
+// watched events on demand. Filtering can be switched using SetDispatchEvents.
+type FilterWatcher struct {
+	watcher informer.Watcher
+
+	dispatchEvents bool
+}
+
+func NewFilterWatcher(config FilterWatcherConfig) (*FilterWatcher, error) {
+	if config.Watcher == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Watcher must not be empty", config)
+	}
+
+	w := &FilterWatcher{
+		watcher: config.Watcher,
+
+		dispatchEvents: true,
+	}
+
+	return w, nil
+}
+
+func (w *FilterWatcher) SetDispatchEvents(dispatchEvents bool) {
+	w.dispatchEvents = dispatchEvents
+}
+
+func (w *FilterWatcher) Watch(listOptions metav1.ListOptions) (watch.Interface, error) {
+	watchInterface, err := w.watcher.Watch(listOptions)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	return watch.Filter(watchInterface, w.filterFunc), nil
```

Comment:
The upstream watch package has this filter implementation which is pretty useful here.

Reply:
Is the cache a global thing there? I'm not deeply familiar our informer implementation. Not sure how apimachinery `watch.Filter` call may affect our custom informer.

Reply:
This here has nothing to do with the cache. The filter watcher is only a gate that forwards to the underlying informer if we want to. In the test we disable the forwarding on purpose and simulate our informer never seeing the delete event and thus never having any chance to purge the deleted runtime object from the cache, which is why the test is failing now.

Reply:
Ack, I understand now.